Added some unit tests for GIN

zhi-yi-huang opened the pull request:
## Description
The GIN algorithm returns the graph and the causal order. In this algorithm, the causal order and the causal graph correspond to each other, so only the causal order in the result is asserted.

## Updates

- Fixed the independence test problem in GIN
- Added assertion to the previous test cases
- Removed the plot function in previous test cases for GIN
- Added tests to test GIN algorithm using the hsic independence test.
- Updated docstring
- Removed default parameter of indep_test
- Refactored the test code

## Test Plan

```python
python -m unittest tests.TestGIN # should pass
```
![image](https://user-images.githubusercontent.com/83860323/180476878-47a93277-d5b7-4688-8591-d1d6e5b280cc.png)

tofuwen commented on tests/TestGIN.py:
```diff
@@ -1,104 +1,81 @@
         data = (data - np.mean(data, axis=0)) / np.std(data, axis=0)
-        g, k = GIN(data)
-        print(g, k)
+        _, k = GIN(data)
```
please give better naming (instead of naming it "k")

Basically you should never use not-meaningful variable name like "k" (only exception is well-known i and j for loops)

zhi-yi-huang replied:
Thanks for the suggestion, I will update it later. I see that k is often used in papers to indicate causal order, e.g. [LiNGAM](https://www.jmlr.org/papers/v7/shimizu06a.html), so I just used k.

tofuwen replied:
hmmm, makes sense haha. But it would still be good to name it clearly in our codebase (you can add comment to say this variable corresponds to k in paper)

tofuwen commented on tests/TestGIN.py:
```diff
@@ -1,104 +1,81 @@
         data = (data - np.mean(data, axis=0)) / np.std(data, axis=0)
-        g, k = GIN(data)
-        print(g, k)
+        _, k = GIN(data)
```
same

tofuwen commented on tests/TestGIN.py:
```diff
@@ -1,104 +1,81 @@
-        g, k = GIN(data)
-        print(g, k)
+        _, k = GIN(data)
+        k = [sorted(k_i) for k_i in k]
```
previously you use i, here you use k_i. 

It would be better if you give consistent naming

zhi-yi-huang replied:
Because 'i' means index, so change the i to k_i, I forgot to unify it when I changed it before.

tofuwen replied:
got it --- please make them consistent

zhi-yi-huang commented on causallearn/search/HiddenCausal/GIN/GIN.py:
```diff
@@ -50,6 +41,11 @@ def GIN(data, indep_test=kci, alpha=0.05):
     n = data.shape[1]
     cov = np.cov(data.T)
 
+    if indep_test_method == 'kci':
+        indep_test = KCI_UInd()
+    else:
+        raise NotImplementedError((f"Independent test method {indep_test_method} is not implemented."))
+
```
This requires kernel-based independence tests such as KCI, HSIC. But they are called differently, and I wonder if I need to write a function to eliminate their differences.

tofuwen replied:
so will the old code support HSIC? Basically my point is that, we should never regress (i.e. previously supported functions are not supported anymore).

if the old code base only supports KCI, then it's fine

zhi-yi-huang commented on causallearn/search/HiddenCausal/GIN/GIN.py:
```diff
@@ -220,7 +212,7 @@ def cal_e_with_gin(data, cov, X, Z):
     _, _, v = np.linalg.svd(cov_m)
     omega = v.T[:, -1]
-    return np.dot(omega, data[:, X].T)
+    return np.dot(data[:, X], omega.T)
```
Reduce the overhead of data matrix transposition.

tofuwen replied:
will this change the function return result? 

The current return value is the old value transpose, right?

zhi-yi-huang replied:
Sorry, omega seems to be a 1-dimensional array and transposing the matrix is an invalid operation. The output result is unchanged. np.dot operation on N-dimensional array with 1-dimensional array is in accordance with the algorithm logic. Please refer to [numpy.dot](https://numpy.org/doc/stable/reference/generated/numpy.dot.html)

zhi-yi-huang commented on tests/TestGIN.py:
```diff
@@ -1,104 +1,81 @@
+        sample_size = 500
+        random.seed(42)
+        np.random.seed(42)
+        L1 = np.random.uniform(-1, 1, size=sample_size)
+        L2 = np.random.uniform(1.2, 1.8) * L1 + np.random.uniform(-1, 1, size=sample_size)
+        X1 = np.random.uniform(1.2, 1.8) * L1 + 0.2 * np.random.uniform(-1, 1, size=sample_size)
+        X2 = np.random.uniform(1.2, 1.8) * L1 + 0.2 * np.random.uniform(-1, 1, size=sample_size)
+        X3 = np.random.uniform(1.2, 1.8) * L2 + 0.2 * np.random.uniform(-1, 1, size=sample_size)
+        X4 = np.random.uniform(1.2, 1.8) * L2 + 0.2 * np.random.uniform(-1, 1, size=sample_size)
```
The previous data generation parameters were set according to the paper, but generating according to the paper does not seem to be fully identifiable. So I changed the data generation parameters.

tofuwen replied:
Why is the old data not fully identifiable? We can prove it not identifiable or just our algorithm fail to do so? Sorry causal n00b lol

zhi-yi-huang replied:
It should be an independence test error caused by the data.

zhi-yi-huang commented:
## Updates

- Added tests to test GIN algorithm using the hsic independence test.
- Updated docstring
- Removed default parameter of indep_test
- Fixed tests

## Test Plan

```python
python -m unittest tests.TestGIN # should pass
```

![image](https://user-images.githubusercontent.com/83860323/179757037-41dc36b3-f7c5-4af8-b067-4d338558ffbd.png)

tofuwen commented on tests/TestGIN.py:
```diff
@@ -1,104 +1,154 @@
+        ground_truth = [[0, 1], [2, 3]]
+        assert len(causal_order) == len(ground_truth)
+        for i in range(len(causal_order)):
+            assert np.isclose(causal_order[i], ground_truth[i]).all()
```
you don't need to use np.isclose() right?

causal_order must equal to ground_truth, right?

zhi-yi-huang replied:
For the first one, I don't understand. For the second one, yes.

tofuwen commented on tests/TestGIN.py:
```diff
@@ -1,104 +1,154 @@
+        for i in range(len(causal_order)):
+            assert np.isclose(causal_order[i], ground_truth[i]).all()
 
+    def test_case1_hsic(self):
```
this function looks almost exactly the same as last function. 

Please consider re-use the code, instead of copy and paste (basically you should never copy and paste code)

You can refer to https://github.com/cmu-phil/causal-learn/pull/59 to make your test clearer.

zhi-yi-huang replied:
Thank you for your suggestion, I will refer to modify the test code.

tofuwen commented on tests/TestGIN.py:
```diff
@@ -1,104 +1,154 @@
+        for i in range(len(causal_order)):
+            assert np.isclose(causal_order[i], ground_truth[i]).all()
 
+    def test_case2_kci(self):
```
same for test case 2.

Basically you can test both kci and hsic in a single function, and the only difference seems to be GIN(data, indep_test_method) indep_test_method this parameter. 

Check https://github.com/cmu-phil/causal-learn/pull/59

tofuwen commented on tests/TestGIN.py:
```diff
@@ -1,104 +1,154 @@
+            assert np.isclose(causal_order[i], ground_truth[i]).all()
+
+
+    def test_case3_kci(self):
```
same

tofuwen reviewed (commented):
Thanks for the great work!

Almost done --- we should do a small refactor to make code more concise. 
Remember the principle: never copy and paste and re-use code whenever possible. :)

tofuwen commented:
BTW, I really like the three tests you designed. I think that's great. :)

Also a small nit: when you made later changes, you can update your description and test plan instead of commenting a new one --- people will generally read the first one instead of scroll down to check all the conversions. This makes this PR clearer for other people. :)

tofuwen commented:
@zhi-yi-huang do you mind addressing the comments? After that, I think we can merge this PR --- very close now

zhi-yi-huang commented:
Sorry, I missed the email from GitHub.

tofuwen commented on tests/TestGIN.py:
```diff
@@ -1,104 +1,88 @@
+    def validate_result(ground_truth, estimated_result):
+        assert len(ground_truth) == len(estimated_result)
+        for i in range(len(estimated_result)):
+            assert np.isclose(estimated_result[i], ground_truth[i]).all()
```
why do we use "isclose()" instead of "==" here? 

With integer comparison, you should expect "==", right?

I think "isclose()" is meant to compare floating point

tofuwen reviewed (commented):
Thanks for the great work, it looks much better now!

Finally one small thing need to be addressed

tofuwen commented:
This is awesome! I think this PR is ready to be merged. :)

cc @kunwuz
